libc/string_catdup.c

```diff
@@ -0,0 +1,749 @@
+	}
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+TEST_SETUP(string_dup)
+{
+}
+
+
+TEST_TEAR_DOWN(string_dup)
+{
+}
+
+
+TEST(string_dup, strdup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_empty)
+{
+	char buf[2] = "";
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_huge_string)
+{
+	char *buf;
+
+	buf = strdup(huge_buf);
+
+	TEST_ASSERT_NOT_NULL(buf);
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+
+	free(buf);
+}
+
+
+TEST(string_dup, strdup_specials_string)
+{
+	char buf[SIZE_BUF * 2] = SPECIALS_STRING;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_cat, strdup_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 1; i < 128; i++) {
+		ascii[i - 1] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_cat, strdup_extended_ascii)
+{
+	int i;
+	char *buf;
+	char ascii[128] = { 0 };
+
+	for (i = 128; i < 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_dup, strdup_non_ascii)
+{
+	char buf[SIZE_BUF * 2] = NON_ASCII_1 NON_ASCII_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+TEST(string_dup, strndup_part)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, strlen(WORD_1));
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_full_string)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, strlen(buf));
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_zero_size)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, 0);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING("", buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_large_size)
+{
+	errno = 0;
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, SIZE_BUF - 1);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_larger_size)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpragmas"
+#pragma GCC diagnostic ignored "-Wstringop-overread"
+
+	errno = 0;
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, SIZE_BIGGER_BUF);
```

Comment:
I think the purpose is similar to the large_size case for strncat, so please try to use SIZE_MAX there and name this case as the strncat one.